### ISSUE TITLE
around destination?

### ISSUE THREAD

**Issue by josharian** [cursorless-dev/cursorless#1799]
Very crude idea here, just jotting it down.

# Problem

Wrapper snippets live in the top level Talon namespace. They compete with all other commands. This makes them hard to name, particularly as they become numerous.

# Proposal

An "around" destination. Then instead of `"if wrap air"` it'd be `"snip if around air"`, just as it is `"snip range before/to/after air"`.

This could also be used as an alternate spoken form for paired delimiter wrapping. Instead of `"curly wrap air"` it could be `"wrap curly around air"`.

This might also open the path to multiple positional holes in snippets, if the target had multiple ranges selected. But this gets complicated quickly.

**Comment by pokey** [comment]
Interesting idea.  How to handle wrapping into different holes of a snippet? Today we use eg `"if wrap air"` vs `"else wrap air"`

**Comment by josharian** [comment]
The obvious analogue is `"snip else around air"`. But maybe I'm missing something more subtle going on there.

**Comment by pokey** [comment]
I guess that could work, it's just that "snip <name>" feels like it is the name of the snippet itself, which is the same snippet no matter which hole you're wrapping into. But I wouldn't say this is a showstopper; I still think this grammar is an interesting idea. Prob worth tossing on the to-discuss list

**Comment by josharian** [comment]
Even deeper into "ponies" land...

`bring` currently serves to provide a kind of ad hoc insertion snippet. If there was some way to textually represent a variable/hole, then it could also serve to provide ad hoc wrapper snippets.

I needed to wrap a dozen select statements today in `coalesce( ... , "unknown")`.

I thought I was clever and created a custom wrapper snippet, but it was a lot of work to get it all hooked up correctly and iron out the bugs, and in the end it was clearly a waste of time. Some text in a split editor like `coalesce(WHAT_GOES_HERE, "unknown")` would have enabled a quick `bring file cap around air past plex` as needed.

**Comment by pokey** [comment]
Yeah ad hoc snippets were one of the eventual ideas behind building the code that we use for "snippet make" today

I also touch briefly on ad hoc snippets in https://github.com/cursorless-dev/cursorless/issues/46

I think nowadays I'd be tempted to do it prefixed by a command like "with hole <target>" to indicate the hole, eg "with hole air bring call air around bat"

**Comment by josharian** [comment]
> How do `"<name> wrap <target>"` vs `"snip <name> around <target>"` differ in this regard?

I thought about this more carefully. The fully ambiguous cases are a bit of a stretch. For example, if `word wrap` toggles word wrap, and you had an insertion snippet called word (why?), then `word wrap air` could be:

* "toggle word wrap" then "insert a"
* "use insertion snippet word with target a"

I'm not entirely convinced that there aren't other less concocted examples involving homophones / near homophones instead of straight up ambiguity.


That said, I think what's really going on is that I mentally model talent commands as a giant decision tree, and I want everything snippet related to be under snip, rather than scattered around as a bunch of early roots. Which is not necessarily a good reason. :)


> Unless the targets are directly adjacent, it's not clear where to put the snippet

Good point.

**Comment by josharian** [comment]
Another use case: the `call` action.

I find it hard to use because it requires selecting the target first (`"take air call bat"`). It could instead be `"call bat around air"`.

**Comment by AndreasArvidsson** [comment]
Actually already supported in our api. You just need to add a spoken form for it. :)

**Comment by josharian** [comment]
Should we do that for everyone? (Do you have a spoken form for it already?)

**Comment by AndreasArvidsson** [comment]
I personally don't use it, but we could do that.

**Comment by pokey** [comment]
We removed it a while ago to improve talon performance. Those performance issues should no longer be a problem, so I'd be happy to have it re-added.

**Comment by AndreasArvidsson** [comment]
I will start on it a asap!

**Comment by AndreasArvidsson** [comment]
pr: https://github.com/cursorless-dev/cursorless/pull/1900
